src/tdsops.f90

```diff
@@ -0,0 +1,700 @@
+      !! corresponding to the tridiagonal system to be solved, and used only in
+      !! the first phase of the distributed algorithm when building the RHS.
+      !! If a boundary condition is defined then coeffs_t and coeffs_b differ
+      !! from coeffs array and define the RHS rule for the first and last 4
+      !! entries in the tridiagonal system (n_halo = 4).
+      !!
+      !! This class does not know about the current rank or its relative
+      !! location among other ranks. All the operator arrays here are used when
+      !! executing a distributed tridiagonal solver phase one or two.
+      real(dp), allocatable, dimension(:) :: dist_fw, dist_bw, dist_sa, dist_sc
+      real(dp), allocatable :: coeffs(:), coeffs_t(:, :), coeffs_b(:, :)
+      integer :: n, n_halo
+      real(dp) :: alpha, delta
+   contains
+      procedure :: deriv_1st, deriv_2nd, interpl_mid, stag_der_1st, preprocess
+   end type tdsops_t
+
+   interface tdsops_t
+      module procedure tdsops
+   end interface tdsops_t
+
+contains
+
+   type(tdsops_t) function tdsops(n, delta, operation, variation, n_halo, &
+                                  from_to, bc_top, bc_bot, paire, c_nu, nu0_nu)
+      !! Constructor function for the tdsops_t class.
+      !! 
+      !! 'n', 'delta', 'operation', and 'variation' are necessary arguments.
+      !! Number of points 'n', distance between two points 'delta', the
+      !! 'operation' the tridiagonal system defines, and the 'variation' that
+      !! specifies the exact scheme we choose to apply for the operation.
+      !! The remaining arguments are optional.
+      !! 'from_to' is necessary for interpolation and staggared derivative, and
+      !! it can be 'v2p' or 'p2v'.
+      !! If the specific region the instance is operating is not a boundary
+      !! region, then 'bc_top' and 'bc_bot' are either 'null' or not provided.
+      !! 'sym' is relevant when the boundary condition is free-slip. If sym is
+      !! .true. then it means the field we operate on is assumed to be an even
+      !! function (symmetric) accross the boundary. If it is .false. it means
+      !! that the field is assumed to be an odd function (anti-symmetric).
+      !! 'c_nu', 'nu0_nu' are relevant when operation is second order
+      !! derivative and variation is compact6-hyperviscous.
+      implicit none
+
+      integer, intent(in) :: n
+      real(dp), intent(in) :: delta
+      character(*), intent(in) :: operation, variation
+      integer, optional, intent(in) :: n_halo
+      character(*), optional, intent(in) :: from_to, bc_top, bc_bot
+      logical, optional, intent(in) :: paire
+      real(dp), optional, intent(in) :: c_nu, nu0_nu
+
+      integer :: n_stencil
+
+      tdsops%n = n
+      tdsops%delta = delta
+      if (present(n_halo)) then
+         tdsops%n_halo = n_halo
+      else
+         tdsops%n_halo = 4
+      end if
+
+      n_stencil = 2*tdsops%n_halo + 1
```

Comment:
I am not sure to understand this line. It leads to an oversized array coeffs in some cases. And the oversized array will lead to useless memory access ?

Reply:
If n_halo is 4 it means RHS at point i is a function of points [i-4, i-3, i-2, i-1, i, i+1, i+2, i+3, i+4]. If halo was 2 then it would be points from i-2 to i+2 so 2*2+1=5 in total.

Reply:
No real impact on performance according to https://github.com/xcompact3d/x3d2/pull/11#discussion_r1369183495